saleor/checkout/fetch.py

```diff
@@ -50,6 +66,157 @@ def get_customer_email(self) -> str:
+        ):
+            from .utils import is_shipping_required
+
+            shipping_required = is_shipping_required(lines)
+        else:
+            shipping_required = checkout_info.checkout.is_shipping_required()
+
+            if not shipping_method or not shipping_required:
+                return zero_taxed_money(checkout_info.checkout.currency)
+
+        shipping_price = shipping_method.channel_listings.get(
+            channel_id=checkout_info.checkout.channel_id,
+        ).get_total()
+
+        return quantize_price(
+            TaxedMoney(net=shipping_price, gross=shipping_price),
+            shipping_price.currency,
+        )
+
+    def is_valid_delivery_method(self) -> bool:
+        if not self.shipping_address:
+            return False
+        return True
+
+    def is_method_in_valid_methods(self, checkout_info: "CheckoutInfo") -> bool:
+        valid_delivery_methods = checkout_info.valid_delivery_methods
+        return bool(
+            valid_delivery_methods and self.delivery_method in valid_delivery_methods
+        )
+
+    def update_channel_listings(self, checkout_info: "CheckoutInfo") -> None:
+        checkout_info.shipping_method_channel_listings = (
+            ShippingMethodChannelListing.objects.filter(
+                shipping_method=self.delivery_method, channel=checkout_info.channel
+            ).first()
+        )
+
+
+@dataclass(frozen=True)
+class CollectionPointInfo(EmptyDeliveryMethod):
+    delivery_method: "Warehouse"
+    shipping_address: Optional["Address"]
+    order_key: str = "collection_point"
+
+    @property
+    def warehouse_pk(self):
+        return self.delivery_method.pk
+
+    @property
+    def is_local_collection_point(self):
+        return (
+            self.delivery_method.click_and_collect_option
+            == WarehouseClickAndCollectOption.LOCAL_STOCK
+        )
+
+    def get_warehouse_filter_lookup(self) -> Dict[str, Any]:
+        return (
+            {"warehouse__pk": self.delivery_method.pk}
+            if self.is_local_collection_point
+            else {}
+        )
+
+    def calculate_checkout_shipping(
+        self, checkout_info: "CheckoutInfo", lines=None
+    ) -> TaxedMoney:
+        return zero_taxed_money(checkout_info.checkout.currency)
+
+    def is_valid_delivery_method(self) -> bool:
+        return (
+            self.shipping_address is not None
+            and self.shipping_address == self.delivery_method.address
+        )
+
+    def is_method_in_valid_methods(self, checkout_info) -> bool:
+        valid_delivery_methods = checkout_info.valid_delivery_methods
+        return bool(
+            valid_delivery_methods and self.delivery_method in valid_delivery_methods
+        )
+
+
+@singledispatch
+def build_delivery_method(
+    delivery_method: Optional[Union["ShippingMethod", "Warehouse"]],
+    address=Optional["Address"],
+) -> EmptyDeliveryMethod:
+    raise NotImplementedError("Incompatible Type")
```

Comment:
Yes, technically it would be the same, I just want to avoid `if else` ladder, and probably, the benefits would be more visible when creating these objects would be more complicated, than just simple `__init__`

Reply:
I have changed to proposed version.